lib/filesystem.nix

```diff
@@ -148,10 +159,131 @@ in
     # The path to recursively list
     dir:
     lib.flatten (lib.mapAttrsToList (name: type:
-    if type == "directory" then
-      lib.filesystem.listFilesRecursive (dir + "/${name}")
-    else
-      dir + "/${name}"
+      if type == "directory" then
+        lib.filesystem.listFilesRecursive (dir + "/${name}")
+      else
+        dir + "/${name}"
   ) (builtins.readDir dir));
 
+  /*
+    Transform a directory tree containing package files suitable for
+    `callPackage` into a matching nested attribute set of derivations.
+
+    For a directory tree like this:
+
+    ```
+    my-packages
+    ├── a.nix
+    ├── b.nix
+    ├── c
+    │  ├── my-extra-feature.patch
+    │  ├── package.nix
+    │  └── support-definitions.nix
+    └── my-namespace
+       ├── d.nix
+       ├── e.nix
+       └── f
+          └── package.nix
+    ```
+
+    `packagesFromDirectory` will produce an attribute set like this:
+
+    ```nix
+    # packagesFromDirectory {
+    #   callPackage = pkgs.callPackage;
+    #   directory = ./my-packages;
+    # }
+    {
+      a = pkgs.callPackage ./my-packages/a.nix { };
+      b = pkgs.callPackage ./my-packages/b.nix { };
+      c = pkgs.callPackage ./my-packages/c/package.nix { };
+      my-namespace = {
+        d = pkgs.callPackage ./my-packages/my-namespace/d.nix { };
+        e = pkgs.callPackage ./my-packages/my-namespace/e.nix { };
+        f = pkgs.callPackage ./my-packages/my-namespace/f/package.nix { };
+      };
+    }
+    ```
+
+    In particular:
+    - `.nix` files are returned with the `.nix` extension removed.
+    - Other files are ignored.
+    - If a directory contains a `package.nix`, the directory name is used as
+      the package name.
+    - Otherwise, directories are recursively traversed.
+
+    Example:
+      packagesFromDirectory {
+        inherit (pkgs) callPackage;
+        directory = ./my-packages;
+      }
+      => { ... }
+
+      lib.makeScope pkgs.newScope (
+        self: packagesFromDirectory {
+          callPackage = self.callPackage;
+          directory = ./my-packages;
+        }
+      )
+      => { ... }
+
+    Type:
+      packagesFromDirectory :: AttrSet -> AttrSet
+  */
+  packagesFromDirectory =
+    # Options.
+    {
+      # `pkgs.callPackage`
+      #
+      # Type:
+      #   (Path | AttrSet -> a) -> AttrSet -> a
+      callPackage
```

Comment:
Just about everyone will want to use `callPackage` here, I don't think users should have to write a `path: pkgs.callPackage path {}` wrapper to use this. For special cases it's just as easy to go the other way (`path: _: import path`).

Reply:
But also I could see adding another argument to provide overrides, maybe bundling this up with `lib.makeScope`:

```nix
packagesFromDirectoryRecursive {
  overrides = self: {
    packageName = /* overrides set */;
  };
}
```

Reply:
Yes, an ergonomic variant of makeScope is good. I'll try to elaborate later.

Reply:
Having talked on Matrix a bit, I do agree that `callPackage` would be nice to have just because it's so well-known, even if it doesn't really make sense for the function as is. So I think having `callPackage` as an argument is fine.

I don't really want to block on this, because this is a useful function, even if it's not perfect.

Reply:
I'll mark this as resolved for now and work on adding `packageFromPath` once we get this merged.